### ISSUE TITLE
Feature Request: improve DX/convinience for IN operator

### ISSUE THREAD

**Issue by moltar** [jawj/zapatos#37]
I think some shortcut for convenience for `IN` operator would be great.

Right now, if I want to use an `IN` operator, I have to drop straight to `SQLFragment` which gets gnarly pretty fast, even for a simple `IN`:

```ts
      db.deletes('foo', {
        id: db.sql<db.SQL>`${db.self} IN (${db.vals(ids)})`,
      })
```

I was wondering, if maybe passing an array of values, would automatically convert it to `IN`?

Example:

```ts
    db.deletes('foo', {
      id: [1, 2, 3],
    })
```

Or maybe a helper?

```ts
    db.deletes('foo', {
      profile_id: db.in([1, 2, 3]),
    })
```

First example would only work for columns with primitive data types.

Second approach could technically made work with any data type, I think.

Thanks!

**Comment by jawj** [comment]
Thanks: I’ll have a think.

**Comment by jawj** [comment]
So I think passing an array is a no-go, since (as you say) that would have to behave differently based on the column type.

The helper approach is fine — but it's so simple to make your own that I wonder if this belongs in the library, or should just be a hint in the documentation?

The helper looks like this:

```typescript
const isIn = (a: any[]) => db.sql<db.SQL>`${db.self} IN (${db.vals(a)})`,
```

And you'd use it like this:

```typescript
const
      authorIds = [1, 2],
      books = await db.select('books', { authorId: isIn(authorIds) }).run(pool);
```

The only drawback of this approach is that the `IN` values aren't type-checked. I think that to type-check them would imply a new interpolation type and a new case in the `SQLFragment` compilation process.

I'm not sure if this is worth the extra complication, but I'm open to arguments either way.

**Comment by moltar** [comment]
> but it's so simple to make your own that I wonder if this belongs in the library

I feel like that's what libraries are for - an opinionated set of tools, and helpers :)

I do realize that maybe a "kitchen sink" of stuff is also not a good idea and it is difficult to find the balance.

But `IN` seems like a such a common pattern, that it makes sense to have it included.

My `db` object isn't global, and is injected via DI. So I am not able to easily create a global helper for it. I could inject a helper too, though.

**Comment by jawj** [comment]
Turns out we can have type-checking without any extra compilation steps, we just have to add an additional generic parameter to `SQLFragment`.

If we're going to support `IN` then there's a bunch of other things we may as well support too. See: https://github.com/jawj/zapatos/blob/condition-helpers/src/conditions.ts

Current plan is to export these under a property, so you would do e.g.:

```typescript
import * as db from './zapatos/src';
const dbc = db.conditions;

const moreBooks = await db.select('books', {
  authorId: dbc.or(dbc.isNotIn(authorIds), dbc.isIn(authorIds)),  // yeah, all of them!
  id: dbc.and(dbc.gt(0), dbc.lte(1000)),
  title: dbc.ne('x'),
  createdAt: dbc.lt(db.sql`NOW()`),
  updatedAt: dbc.isNotNull,
}).run(pool);
```

**Comment by jawj** [comment]
One of the reasons it was looking gnarly without the new helpers was the unwanted type inference on the `Interpolations` generic when using `sql`. I think I've fixed this, so that:

```typescript
db.deletes('foo', { id: db.sql<db.SQL>`${db.self} IN (${db.vals(ids)})` })
```

loses the `<db.SQL>` to become:

```typescript
db.deletes('foo', { id: db.sql`${db.self} IN (${db.vals(ids)})` })
```

Between this and the condition helpers, there are couple of small but deep changes to the core types. Would you mind trying this out and kicking the tyres before I release to npm?

I think you just need to use `npm install jawj/zapatos#condition-helpers` (not forgetting to then regenerate your `zapatos` folder).

**Comment by jawj** [comment]
OK, I needed to make a release to fix some other issues (#38), so I've released these changes too in 0.1.57. Please shout if anything breaks.

**Comment by moltar** [comment]
@jawj thank you so much for this! Usage is now so clean. And sorry for the delay in testing, only got around to it today. Everything I've tested so far works great. Cheers!
